File: misc/advent/2020/15/a.zig
const std = @import("std");
const fs = std.fs;

const LastSaid = std.AutoHashMap(usize, usize);

pub fn solve(inp: []usize, stop: usize, alloc: *std.mem.Allocator) usize {
    var said = LastSaid.init(alloc);
    defer said.deinit();

    // factor of 64 determined empirically
    said.ensureCapacity(@intCast(u32, stop / 64)) catch unreachable;

    for (inp[0 .. inp.len - 1]) |elt, i| {
        said.put(elt, i + 1) catch unreachable;
    }

    var turn: usize = inp.len + 1;
    var current: usize = inp[inp.len - 1];
    while (true) {
        var res = said.getOrPut(current) catch unreachable;
        if (res.found_existing) {
            current = (turn - 1) - res.entry.value;
        } else {
            current = 0;
        }
        res.entry.value = turn - 1;
        turn += 1;
        if (turn > stop) {
            return current;
        }
    }
}

test "solve" {
    var alloc = std.heap.GeneralPurposeAllocator(.{}){};
    var inp = [_]usize{ 0, 3, 6 };
    std.testing.expect(solve(inp[0..], 2020, &alloc.allocator) == 436);
}

pub fn main() !void {
    var alloc = std.heap.GeneralPurposeAllocator(.{}){};
    var inp = [_]usize{ 1, 0, 18, 10, 19, 6 };
    std.debug.print("{}\n", .{solve(inp[0..], 2020, &alloc.allocator)});
    std.debug.print("{}\n", .{solve(inp[0..], 30_000_000, &alloc.allocator)});
}

File: misc/advent/2020/15/a.py
from collections import defaultdict


def run(inp, stop=2020, verbose=False):
    said = {j: i + 1 for i, j in enumerate(inp[:-1])}

    turn = len(inp) + 1
    current = inp[-1]
    while 1:
        if current not in said:
            if verbose:
                print(f"{turn}: 0 (adding {current} {dict(said.items())})")
            said[current] = turn - 1
            current = 0
        else:
            if verbose:
                print(
                    f"{turn}: {(turn-1) - said[current][0]} ({turn-1} - {said[current][0]}) {current} {dict(said.items())}"
                )
            c = (turn - 1) - said[current]
            said[current] = turn - 1
            current = c
        turn += 1
        if turn > stop:
            return current


assert run([1,3,2]) == 1
assert run([2,1,3]) == 10
assert run([1,2,3]) == 27
assert run([2,3,1]) == 78
assert run([3,2,1]) == 438
assert run([3,1,2]) == 1836

# assert run([0,3,6], stop=30000000) == 175594
print(run([1, 0, 18, 10, 19, 6], stop=2020))
print(run([1, 0, 18, 10, 19, 6], stop=30000000))
